### ISSUE TITLE
Add Dos MZ support to retdec

### ISSUE THREAD

**Issue by DerJJ** [avast-tl/retdec#59]
Being able to decompile old MS DOS/MZ programs would be nice.

**Comment by asciibox** [comment]
Yes, I'm struggling with this right now too and also would upvote this

**Comment by marespiaut** [comment]
Hello,
I also would like to use RetDec to decompile MS-DOS binaries.
While it may not be useful for modern security purpose, it would help all the hobbyist around here trying to reverse engineer MS-DOS video games.

Best regards.

**Comment by PeterMatula** [comment]
If anyone wants to help with this one, here are some pointers what needs to be done (and in what order):
1. Add Dos MZ parser to `fileformat`.
2. Add Dos MZ loader to `loader`.
3. Modify scripts and other support code to let Dos MZ through.
4. Modify `decoder` pass in `bin2llvmir` to handle 16-bit x86. 
5. As we wrote `capstone2llvmir`, we tried to model 16-bit x86 alongside 32/64-bit. But we have never used it so far, so it is quite possible some problems will need to be fixed, and some more features added.

At this point, decompilation of 16-bit x86 Dos MZ file should produce C. But the output quality probably will not be very good. So:

5. Pick some very simple binary, decompile it, inspect the results, find out what is wrong, find out why it is wrong, draft a solution, implement it. (E.g. the whole x86 segment registers thing is not dealt with at all, DOS interrupts, etc. Either existing `retdec` LLVM passes will need to be modified, or new, DOS specific passes will have to be created.)
6. Write regression tests when results are ok.
7. Repeat points 5. and 6. with some more complex binaries.

**Comment by marespiaut** [comment]
Hello @PeterMatula,
Thank you very much for the details. I'll definitely give it a try when I'll have some free time.
However, I'm pretty sure 90% of us want to decompile 32-bit x86 Dos MZ, and not 16-bit x86.

**Comment by greenozon** [comment]
Are there any news after 5 years regarding MS DOS 16 bit binaries support?
